oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,569 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Data structure for expressing quantile estimation.
+//! This is based on the P² heuristic algorithm for dynamic
+//! calculation of the median and other quantiles. The estimates
+//! are produced dynamically as the observations are generated.
+//! The observations are not stored; therefore, the algorithm has
+//! a very small and fixed storage requirement regardless of the
+//! number of observations.
+//!
+//! Read the [paper](https://www.cs.wustl.edu/~jain/papers/ftp/psqr.pdf)
+//! for more specifics.
+
+// Copyright 2024 Oxide Computer Company
+
+use crate::traits::HistogramSupport;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use thiserror::Error;
+
+const FILLED_MARKER_LEN: usize = 5;
+
+/// Errors related to constructing a `Quantile` instance or estimating the
+/// p-quantile.
+#[derive(
+    Debug, Clone, Error, JsonSchema, Serialize, Deserialize, PartialEq,
+)]
+#[serde(tag = "type", content = "content", rename_all = "snake_case")]
+pub enum QuantileError {
+    /// The p value must be in the range [0, 1].
```

Comment:
Just checking, does the algorithm work correctly if someone passes `0.0` or `1.0` to the constructor? That is, can we actually "estimate" the minimum and maximum? The paper says:

> For the algorithm to work correctly, the marker heights must always be in a nondecreasing order

That implies it _should_ work to estimate the 0 or 100th quantile, but they also don't give examples that I can see. Those values are also sometimes not considered "true" quantiles, so we might want to disallow them anyway. Up to you, but something to check.

Reply:
It should, but I'll double check these bounds.

Reply:
tests added for these and one with a comparison to what another impl returns at 1.0.